test/data/test_modules.py

```diff
@@ -0,0 +1,103 @@
+import torch
+from torchtext.modules import MultiheadAttentionContainer, \
+    ScaledDotProduct, MultiheadInProject, MultiheadOutProject
+from torch.nn.functional import multi_head_attention_forward as mha_forward
+from torch.testing import assert_allclose
+from ..common.torchtext_test_case import TorchtextTestCase
+
+
+class TestModels(TorchtextTestCase):
+
+    def test_multiheadattention(self):
+        embed_dim, nhead, tgt_len, src_len, bsz = 10, 5, 6, 10, 64
+        # Build torchtext MultiheadAttention module
+        MHA = MultiheadAttentionContainer((MultiheadInProject(embed_dim, nhead),
+                                          MultiheadInProject(embed_dim, nhead),
+                                          MultiheadInProject(embed_dim, nhead)),
+                                          ScaledDotProduct(),
+                                          MultiheadOutProject(embed_dim // nhead, nhead))
+
+        query = torch.rand((tgt_len, bsz, embed_dim))
+        key = value = torch.rand((src_len, bsz, embed_dim))
+        attn_mask_2D = torch.randint(0, 2, (tgt_len, src_len)).to(torch.bool)
+        bias_k = bias_v = torch.rand((1, 1, embed_dim))
+        mha_output, attn_weights = MHA(query, key, value,
+                                       attn_mask=torch.stack([attn_mask_2D] * (bsz * nhead)),
+                                       bias_k=bias_k.repeat(1, bsz, 1).reshape(1, bsz * nhead, -1),
+                                       bias_v=bias_v.repeat(1, bsz, 1).reshape(1, bsz * nhead, -1))
+
+        # Use torch.nn.functional.multi_head_attention_forward
+        torch_attn_mask = torch.zeros((tgt_len, src_len)).masked_fill_(attn_mask_2D, float('-inf'))
+        in_proj_weight = torch.cat([MHA.query_in_proj.proj_layer.weight,
+                                    MHA.key_in_proj.proj_layer.weight,
+                                    MHA.value_in_proj.proj_layer.weight])
+        torch_mha_output, torch_mha_weights = mha_forward(query, key, value,
+                                                          embed_dim, nhead,
+                                                          in_proj_weight, None,
+                                                          bias_k, bias_v,
+                                                          False, 0.0,
+                                                          MHA.out_proj.proj_layer.weight,
+                                                          MHA.out_proj.proj_layer.bias,
+                                                          attn_mask=torch_attn_mask)
+
+        assert_allclose(mha_output, torch_mha_output)
+        # With bias_k and bias_v, src_len needs to plus 1
+        attn_weights = attn_weights.view(bsz, nhead, tgt_len, src_len + 1).sum(dim=1) / nhead
+        assert_allclose(attn_weights, torch_mha_weights)
+
+    def test_broadcast_scaled_dot_product(self):
```

Comment:
I'd add some "self.assertRaises" to exercises some explicit cases where we expect broadcasting to fail definitely.